java/arcs/core/storage/ReferenceModeStore.kt

```diff
@@ -208,6 +209,7 @@ class ReferenceModeStore private constructor(
     ): Boolean {
         log.verbose { "onProxyMessage: $message" }
         val refModeMessage = message.sanitizeForRefModeStore(type)
+        devToolsProxy?.onRefModeStoreProxyMessage(message)
```

Comment:
It'd be ideal to not have to require store implementations to be aware of the dev tools system at all. Is there a way to invert these dependencies? 

As it is right now, it's a bit invasive to add devTools support to the system, and it's quite narrowly-scoped to just the dev tools use case. If we absolutely have to adjust the APIs for the storage stack to accomodate stuff like the dev tools system, I'd prefer those API adjustments to be generic/abstract and not be bespoke for dev tooling.

To re-state, I think there are at least three ways to go about doing this (ignoring the details for now):

1) Inject the dev tools proxies directly into the stores as you've got now, modifying store implementation to be aware of the dev tools infra and making it their responsibility to call that stuff themselves.
2) Inject some kind of abstract interface into the stores, not bound to the dev tooling stuff directly (just that dev tools is _an_ implementer) - still requires making it the responsibility of the stores to call into this interface correctly.
3) Utilize existing eventing infrastructure (like proxycallbacks) and such, and make it the dev tooling infra's responsibility to know where to hook up the listeners on any stores its given.

While (3) is ideal in my opinion, it may not be powerful-enough for what you want to achieve.... so to that end, maybe there's an option 2.5 - somewhere between (2) and (3).

Reply:
I do understand where you are coming from and I agree that less invasive solution would be preferable if possible. However I still believe this direction (option (1) - direct integration) is the right approach for what we want to achieve.

WRT option (2): My concern is that making the devtool to storage stack (or arc host) communication abstract and seemingly not related to DevTools will do more harm than good, as I cannot come up with other component needing the type of instrumentation we want to do. In particular we want a 2-way communication between the devtools and various Arcs classes (e.g. please list all storage keys, please output the content of the storage key X), which I believe will be unique to dev tooling.

WRT option (3): There is a lot of information that we can reconstruct from proxy callbacks events, but as you noted yourself, it's not enough for what we want to do. A canonical example is retrieving data that's not yet used in any Arc directly from the driver - this cannot be possibly done in any other way than allowing duplex communication between dev tooling and the exact classes implementing relevant functionality - here, the storage drivers.

I believe that for the scope of what we're aiming for we need to have direct instrumentation of Arcs classes, be it the storage stack, the arc hosts or other relevant pieces.

I have expected this kind of push back, and it's an important conversation we need to go through. CC @shans as well, so that we're all on the same page.

Reply:
Let's keep it how it is for now. If we have to change it later it's no biggie.